components/global/core/io.cellery.observability.telemetry.deduplicator/src/main/java/io/cellery/observability/telemetry/deduplicator/DeduplicationStreamProcessor.java

```diff
@@ -0,0 +1,391 @@
+
+        String currentEventTraceId = (String) traceIdExecutor.execute(currentEvent);
+        String currentEventSpanId = (String) spanIdExecutor.execute(currentEvent);
+        String currentEventParentSpanId = (String) parentSpanIdExecutor.execute(currentEvent);
+        String currentEventSourceNamespace = (String) sourceNamespaceExecutor.execute(currentEvent);
+        String currentEventSourceInstance = (String) sourceInstanceExecutor.execute(currentEvent);
+        String currentEventSourceComponent = (String) sourceComponentExecutor.execute(currentEvent);
+        String currentEventDestinationNamespace = (String) destinationNamespaceExecutor.execute(currentEvent);
+        String currentEventDestinationInstance = (String) destinationInstanceExecutor.execute(currentEvent);
+        String currentEventDestinationComponent = (String) destinationComponentExecutor.execute(currentEvent);
+
+        return Objects.equals(currentEventSourceNamespace, oldEventSourceNamespace)
+                && Objects.equals(currentEventSourceInstance, oldEventSourceInstance)
+                && Objects.equals(currentEventSourceComponent, oldEventSourceComponent)
+                && Objects.equals(currentEventDestinationNamespace, oldEventDestinationNamespace)
+                && Objects.equals(currentEventDestinationInstance, oldEventDestinationInstance)
+                && Objects.equals(currentEventDestinationComponent, oldEventDestinationComponent)
+                && Objects.equals(currentEventTraceId, oldEventTraceId)
+                && (Objects.equals(currentEventParentSpanId, oldEventSpanId)
+                || Objects.equals(currentEventSpanId, oldEventParentSpanId));
+    }
+
+    private StreamEvent mergeEvents(StreamEvent oldEvent, StreamEvent currentEvent) {
+        Long currentEventRequestSizeBytes = (Long) requestSizeExecutor.execute(currentEvent);
+        Long currentEventResponseDuration = (Long) responseDurationExecutor.execute(currentEvent);
+        Long currentEventResponseSizeBytes = (Long) responseSizeExecutor.execute(currentEvent);
+        Long maxRequestSizeBytes = currentEventRequestSizeBytes;
+        Long maxResponseDuration = currentEventResponseDuration;
+        Long maxResponseSizeBytes = currentEventResponseSizeBytes;
+
+        if (oldEvent != null) {
+            Long oldEventRequestSizeBytes = (Long) requestSizeExecutor.execute(oldEvent);
+            Long oldEventResponseDuration = (Long) responseDurationExecutor.execute(oldEvent);
+            Long oldEventResponseSizeBytes = (Long) responseSizeExecutor.execute(oldEvent);
+            maxRequestSizeBytes = Long.max(oldEventRequestSizeBytes, currentEventRequestSizeBytes);
+            maxResponseDuration = Long.max(oldEventResponseDuration, currentEventResponseDuration);
+            maxResponseSizeBytes = Long.max(oldEventResponseSizeBytes, currentEventResponseSizeBytes);
+        }
+
+        Object[] currentOutputs = currentEvent.getOutputData();
```

Comment:
Let's use the `populateComplexEvent()` method to do this without modifying the array on our own.

https://github.com/wso2/cellery-observability/blob/v0.5.0/components/global/core/io.cellery.observability.k8s.client/src/main/java/io/cellery/observability/k8s/client/GetComponentPodsStreamProcessor.java#L188